src/proxy.c

```diff
@@ -225,124 +222,125 @@ cc_proxy_disconnect (struct cc_proxy *proxy)
-	 */
-	bytes_read = recvmsg(proxy_data->socket_fd, &msg, 0);
+	while (1) {
+		bytes_read = recvmsg(proxy_fd, &msg, 0);
+		if (bytes_read < 0 && (errno == EAGAIN || errno == EINTR)) {
+			continue;
+		}
+		break;
+	}
 
 	if (bytes_read == -1){
-		g_critical("failed to read proxy message %s", strerror(errno));
+		g_critical("recvmsg failed: %s", strerror(errno));
 		return false;
 	}
 
 	if (bytes_read == 0) {
-		g_critical("failed to read length");
+		g_critical("recvmsg failed: EOF");
 		return false;
 	}
 
-	total_bytes_read += bytes_read;
+	if (bytes_read != 1 || iov_buffer[0] != OOB_FD_FLAG) {
+		g_critical("recvmsg failed: read %zd bytes, flag: %c",
+			   bytes_read, iov_buffer[0]);
+		return false;
+	}
 
-	if (proxy_data->oob_fd && !proxy_data->oob_fd_found) {
-		/*
-		 * We expect OOB data that we have not found yet.
-		 * The OOB message must come first, as a 1 byte
-		 * only message.
-		 */
-		cmsg = CMSG_FIRSTHDR(&msg);
-		if (!cmsg) {
-			g_critical("could not read the control message");
-			goto out;
-		}
+	cmsg = CMSG_FIRSTHDR(&msg);
+	if (!cmsg) {
+		g_critical("could not read the control message");
+		return false;
+	}
 
-		data = CMSG_DATA(cmsg);
-		if (! (data && iov_buffer[0] == OOB_FD_FLAG)) {
-			g_critical("missing out of band data");
-			goto out;
-		}
+	data = CMSG_DATA(cmsg);
+	if (!data) {
+		g_critical("missing out of band data");
+		return false;
+	}
 
-		*proxy_data->oob_fd = *((int*) data);
-		g_message("oob fd read from proxy socket %d", *proxy_data->oob_fd);
-		proxy_data->oob_fd_found = true;
+	*fd = *((int*) data);
 
-		/* Kick this routine again for the actual command reply */
-		return true;
-	}
+	g_message("received fd from proxy %d", *fd);
 
-	/* Get the proxy message length */
-	memcpy (header_length, iov_buffer, HEADER_MESSAGE_LENGTH);
-	payload_length = cc_oci_get_big_endian_32 (header_length);
-	g_debug ("msg length: %ld", payload_length);
+	return true;
+}
 
-	/* Copy the rest of the io buffer */
-	g_string_append_len (proxy_data->msg_received,
-			iov_buffer + MESSAGE_HEADER_LENGTH,
-			(ssize_t)(bytes_read - MESSAGE_HEADER_LENGTH));
+/**
+ * Read a message from proxy's socket.
+ *
+ * \param source GIOChannel.
+ * \param condition GIOCondition.
+ * \param proxy_data struct watcher_proxy_data.
+ *
+ * \return \c true on success, else \c false.
+ */
+static gboolean
+cc_proxy_msg_read (GIOChannel *source, GIOCondition condition,
+	struct watcher_proxy_data *proxy_data)
+{
+	ssize_t bytes_read, total_bytes_read = 0;
+	size_t payload_length;
+	guint8 header[MESSAGE_HEADER_LENGTH] = { 0 };
+	gchar buf[LINE_MAX];
 
-	/* The entire message was read */
-	if (bytes_read >= payload_length + MESSAGE_HEADER_LENGTH) {
+	/*
+	 * Start by reading the message header and the payload length.
+	 */
+	bytes_read = read(proxy_data->socket_fd, header, sizeof(header));
+	if (bytes_read <= 0) {
+		g_critical("couldn't read header from proxy: %s",
+			   strerror(errno));
 		goto out;
```

Comment:
I've a feeling `goto out` now goes to some code (the msg_received->len check code) that is redundant, as at this point `len` will always be 0. That appears to be the only jump to out:, so maybe it can be moved further down at the bottom of the function? (unless len can be incremented somewhere else I have missed?)